sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -472,6 +525,75 @@ void Scheduler::cleanupCommands(const std::vector<Command *> &Cmds) {
   }
 }
 
+void Scheduler::deferMemObjRelease(const std::shared_ptr<SYCLMemObjI> &MemObj) {
+  {
+    std::lock_guard<std::mutex> Lock{MDeferredMemReleaseMutex};
+    MDeferredMemObjRelease.push_back(MemObj);
+  }
+  cleanupDeferredMemObjects(false);
+}
+
+inline bool Scheduler::isNoDeferredMemObjects() {
+  std::lock_guard<std::mutex> Lock{MDeferredMemReleaseMutex};
+  return MDeferredMemObjRelease.empty();
+}
+
+void Scheduler::cleanupDeferredMemObjects(bool ForceWait) {
+  if (isNoDeferredMemObjects())
+    return;
+
+  // Need to aggregate ready to release object to acquire write lock once.
+  std::list<std::shared_ptr<SYCLMemObjI>> ObjsReadyToRelease;
+  {
+    ReadLockT Lock(MGraphLock, std::try_to_lock);
+    // if we need blocking mode - force lock waiting
+    if (!Lock.owns_lock() && ForceWait)
+      Lock.lock();
+    if (Lock.owns_lock()) {
+      // Not expected that ForceWait == true with be used in parallel with
+      // adding MemObj to storage, no such scenario.
+      std::lock_guard<std::mutex> LockDef{MDeferredMemReleaseMutex};
+      auto MemObjIt = MDeferredMemObjRelease.begin();
+      while (MemObjIt != MDeferredMemObjRelease.end()) {
+        MemObjRecord *Record = MGraphBuilder.getMemObjRecord((*MemObjIt).get());
+        if (!Record) {
+          // Just trigger delete since no operations on object was perfromed and
```

Comment:
As an option, we could avoid adding a mem object to the list without any `Record` in the first place( in `deferMemObjRelease`).

Reply:
Yes, I made it intentionally as it is, because to obtain record we need to:
1) obtain access to scheduler
2) lock graph in shared mode
So I tried to do this as centralized as possible.

Reply:
after discussion fixed in https://github.com/intel/llvm/pull/6837/commits/1e75448f8dcc41d6c2fe51fbfe32fa93066562f0